code/atom/UpdateOverlays.dm

```diff
@@ -140,63 +161,132 @@ ClearSpecificOverlays(1, "key0", "key1", "key2") 	//Same as above but retains ca
 	#if DM_VERSION < 515 && !defined(SPACEMAN_DMM)
 	if(I)
 		for(var/ikey in overlay_refs)
-			if(ikey != key && overlay_refs[ikey][P_INDEX] > 0 && overlay_refs[ikey][P_ISTATE] == ref(I.appearance))
+			if(ikey != key && overlay_refs[ikey][P_INDEX] && overlay_refs[ikey][P_ISTATE] == hash)
 				// logTheThing(LOG_DEBUG, null, "Attempt to add duplicate overlay appearances on [identify_object(src)] with keys [key] and [ikey].")
 				I.layer += 0.0000001 * rand()
 	#endif
 
 	var/index = prev_data[P_INDEX]
-	if(index > 0) //There is an existing overlay in place in this slot, remove it
+	if(index) //There is an existing overlay in place in this slot, remove it
 		if(index <= length(src.overlays))
 			src.overlays.Cut(index, index+1) //Fuck yoooou byond (this gotta be by index or it'll fail if the same thing's in overlays several times)
 		else
 			stack_trace("Overlays on [identify_object(src)] were modified by non-UpdateOverlays method.")
 
-		prev_data[P_INDEX] = 0
 		for(var/ikey in overlay_refs) //Because we're storing the position of each overlay in the list we need to shift our indices down to stay synched
 			var/list/L = overlay_refs[ikey]
-			if(L?.len > 0 && L[P_INDEX] >= index)
+			if(L?[P_INDEX] >= index)
 				L[P_INDEX]--
 
 	if(I)
 		src.overlays += I
-		index = length(src.overlays)
-		prev_data[P_INDEX] = index
+		prev_data[P_INDEX] = length(src.overlays)
 
 		prev_data[P_IMAGE] = I
-		prev_data[P_ISTATE] = "\ref[I.appearance]"
+		prev_data[P_ISTATE] = hash
 
 		overlay_refs[key] = prev_data
 	else
 		if(retain_cache) //Keep the cached image available?
 			prev_data[P_INDEX] = 0	//Clear the index
-			prev_data[P_ISTATE] = 0	//Clear the ref
+			prev_data[P_ISTATE] = null	//Clear the ref
 		else
 			overlay_refs -= key
 	return 1
 
+/atom/proc/AddOverlays(var/image/I, var/key, var/force=0)
+	if(isnull(key))
+		CRASH("AddOverlays called without a key.")
+	LAZYLISTINIT(src.overlay_refs)
+	ASSERT(I)
+
+	var/list/prev_data
+	//List to store info about the last state of the icon
+	prev_data = overlay_refs[key]
+	if(isnull(prev_data)) //Ok, we don't have previous data, but we will add an overlay
+		prev_data = new /list(P_ILEN)
+
+	var/hash = ref(I.appearance)
+	var/image/prev_overlay = prev_data[P_IMAGE] //overlay_refs[key]
+	if(!force && (prev_overlay == I) && hash == prev_data[P_ISTATE]) //If it's the same image as the other one and the appearances match then do not update
+		return 0
+
+	#if DM_VERSION > 515
+	#warn "okay now remove this"
+	#endif
+	#if DM_VERSION < 515 && !defined(SPACEMAN_DMM)
+	for(var/ikey in overlay_refs)
+		if(ikey != key && overlay_refs[ikey][P_INDEX] && overlay_refs[ikey][P_ISTATE] == hash)
+			// logTheThing(LOG_DEBUG, null, "Attempt to add duplicate overlay appearances on [identify_object(src)] with keys [key] and [ikey].")
+			I.layer += 0.0000001 * rand()
+	#endif
```

Comment:
we don't support below 515 anymore, this can be removed

Reply:
otherwise looks good, if others want to review you can limit the commits to look at

Reply:
done